[AMDGPU][Attributor] Make `AAAMDFlatWorkGroupSize` honor existing attribute

shiltian opened the pull request:
If a function has `amdgpu-flat-work-group-size`, honor it in `initialize` by
taking its value directly; otherwise, it uses the default range as a starting
point. We will no longer manipulate the known range, which can cause issues
because the known range is a "throttle" to the assumed range such that the
assumed range can't get widened properly in `updateImpl` if the known range is
not set properly for whatever reasons. Another benefit of not touching the known
range is, if we indicate pessimistic state, it also invalidates the AA such that
`manifest` will not be called. Since we honor the attribute, we don't want and
will not add any half-baked attribute added to a function.

shiltian commented:
* **#114357** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114357?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -728,12 +735,43 @@ struct AAAMDSizeRangeAttribute
+                           /* ForceReplace=*/true);
+  }
+
   ChangeStatus emitAttributeIfNotDefault(Attributor &A, unsigned Min,
```
This will be eventually removed.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -767,11 +805,17 @@ struct AAAMDFlatWorkGroupSize : public AAAMDSizeRangeAttribute {
+    if (Attr.has_value()) {
+      auto [Min, Max] = *Attr;
+      intersectKnown(ConstantRange(APInt(32, Min), APInt(32, Max + 1)));
+      indicatePessimisticFixpoint();
```
This pessimistic fixed point will not make the AA invalid because the `known -> assumed` range is no longer a full set.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -728,12 +735,43 @@ struct AAAMDSizeRangeAttribute
 
     bool AllCallSitesKnown = true;
-    if (!A.checkForAllCallSites(CheckCallSite, *this, true, AllCallSitesKnown))
+    if (!A.checkForAllCallSites(CheckCallSite, *this,
+                                /*RequireAllCallSites=*/true,
+                                AllCallSitesKnown))
       return indicatePessimisticFixpoint();
```
This will make the AA invalid as well because if `updateImpl` is called, the known range is full set for sure, and `known -> assumed` will make the assumed range a full set, thus in an invalid state.

arsenm replied:
Precommit NFC change?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -168,7 +168,15 @@ class AMDGPUInformationCache : public InformationCache {
-  std::pair<unsigned, unsigned> getFlatWorkGroupSizes(const Function &F) {
+  std::optional<std::pair<unsigned, unsigned>>
+  getFlatWorkGroupSizeAttr(const Function &F) const {
+    Attribute A = F.getFnAttribute("amdgpu-flat-work-group-size");
+    if (!A.isStringAttribute())
+      return std::nullopt;
```
The presence or absence of an attribute should not impart meaning. A function explicitly marked with the default value should be the same as one with no attribute.

This will also miss applying the calling convention based default range

shiltian replied:
> A function explicitly marked with the default value should be the same as one with no attribute.

I think we still need to honor existing value, no matter whether it is default or not. FWIW, with this change, at least our compiler will not add default value to the attribute. If a default value appears in the attribute, it must be from users.

shiltian replied:
> The presence or absence of an attribute should not impart meaning. A function explicitly marked with the default value should be the same as one with no attribute.

Give what the front end is doing, we will indeed discard the default attribute.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -767,11 +805,17 @@ struct AAAMDFlatWorkGroupSize : public AAAMDSizeRangeAttribute {
+    std::optional<std::pair<unsigned, unsigned>> Attr =
+        InfoCache.getFlatWorkGroupSizeAttr(*F);
+
+    if (Attr.has_value()) {
```
```suggestion
    if (Attr) {
```

arsenm reviewed (commented):
Test changes?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -168,11 +168,21 @@ class AMDGPUInformationCache : public InformationCache {
+  std::optional<std::pair<unsigned, unsigned>>
+  getFlatWorkGroupSizeAttr(const Function &F) const {
+    Attribute A = F.getFnAttribute("amdgpu-flat-work-group-size");
+    if (!A.isStringAttribute())
```
This amounts to querying the attribute twice. If you're going to apply your own maximum clamping later, can you directly parse the attribute instead of going through the smarter wrapper applying the clamp? i.e. use getIntegerPairAttribute? This would be easier to follow and you can avoid the second map lookup

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -144,6 +144,20 @@ static bool funcRequiresHostcallPtr(const Function &F) {
 
 namespace {
+
+std::optional<std::pair<unsigned, unsigned>> static parseRangeAttribute(
```
Can you add an overload of getIntegerPairAttribute that takes the raw Attribute instead of duplicating the parsing? This also omits the context errors (I guess it's also bad that we'll spam them everywhere. We should probably make this an IR verifier check)

choikwa commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -767,13 +802,33 @@ struct AAAMDFlatWorkGroupSize : public AAAMDSizeRangeAttribute {
+
+    if (auto Attr = InfoCache.getFlatWorkGroupSizeAttr(*F)) {
+      // We will only consider an attribute different from max because the front
+      // end always emits the attribure, unfortunately, and sometimes it emits
```
typo

choikwa commented on llvm/test/CodeGen/AMDGPU/inline-attr.ll:
```diff
@@ -6,14 +6,14 @@
-; NOINFS: attributes #0 = { nounwind "amdgpu-waves-per-eu"="4,10" "no-infs-fp-math"="true" "uniform-work-group-size"="false" }
-; NOINFS: attributes #1 = { nounwind "less-precise-fpmad"="false" "no-infs-fp-math"="true" "no-nans-fp-math"="false" "uniform-work-group-size"="false" "unsafe-fp-math"="false" }
+; NOINFS: attributes #0 = { nounwind "no-infs-fp-math"="true" "uniform-work-group-size"="false" "unsafe-fp-math"="true" }
+; NOINFS: attributes #1 = { nounwind "less-precise-fpmad"="false" "no-infs-fp-math"="true" "no-nans-fp-math"="false" "uniform-work-group-size"="false" "unsafe-fp-math"="true" }
```
Should this change "unsafe-fp-math"="true"?

shiltian commented:
bump

choikwa commented:
LGTM

shiltian commented:
@choikwa you might forget to accept it.

shiltian commented:
ping, I prefer not to update tests again and again...

shiltian committed:
[AMDGPU][Attributor] Make `AAAMDFlatWorkGroupSize` honor existing attribute

If a function has `amdgpu-flat-work-group-size`, honor it in `initialize` by
taking its value directly; otherwise, it uses the default range as a starting
point. We will no longer manipulate the known range, which can cause issues
because the known range is a "throttle" to the assumed range such that the
assumed range can't get widened properly in `updateImpl` if the known range is
not set properly for whatever reasons. Another benefit of not touching the known
range is, if we indicate pessimistic state, it also invalidates the AA such that
`manifest` will not be called. Since we honor the attribute, we don't want and
will not add any half-baked attribute added to a function.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -168,9 +168,18 @@ class AMDGPUInformationCache : public InformationCache {
+    auto R = AMDGPU::getIntegerPairAttribute(F, "amdgpu-flat-work-group-size");
+    if (!R)
+      return std::nullopt;
+    return std::make_pair(R->first, *(R->second));
```
Extra parens?

arsenm commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h:
```diff
@@ -936,6 +936,19 @@ getIntegerPairAttribute(const Function &F, StringRef Name,
+/// \returns \p std::nullopt and emits error if one of the requested values
+/// cannot be converted to integer, or \p OnlyFirstRequired is false and
+/// "second" value is not present.
+std::optional<std::pair<unsigned, std::optional<unsigned>>>
```
Not sure why the second entry is optional. This is spreading a parsing detail to users of the value. Users should be able to just treat it as-if it always has 2 components, the one entry forms are just a compatibility thing with old IR

arsenm replied:
That should never happen. But as a parsing function, that should be abstracted from the attributor's usage

jtramm commented:
Looks like this PR resulted in about a 2x slowdown on MI250 for the most expensive kernel in OpenMC, resulting in an overall 20% performance degradation. Any ideas what might have caused it?

shiltian commented:
You might have device-libs that still come with half-baked values. To verify it, you could use `llvm-dis` to convert it to `.ll` file, remove all attributes `amdgpu-flat-work-group-size`, convert it back to `.bc` file, and put it back. After that, compile everything of OpenMC.

jtramm commented:
Ah, interesting! Which file(s) should I be converting to `.ll`? And how do I remove attributes in a `.ll` file -- just delete the entire line or change to something else? Any specific commands to give to `llvm-dis` for converting between file types?

jtramm commented:
Here's what my current cmake line looks like for LLVM:

```
PROJECTS="clang;lld"
RUNTIMES="openmp;offload"
TARGETS="all"
USE_CCACHE=OFF

cmake \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DLLVM_ENABLE_PROJECTS=${PROJECTS}  \
    -DLLVM_ENABLE_RUNTIMES=${RUNTIMES}   \
    -DLLVM_TARGETS_TO_BUILD=${TARGETS}   \
    -DLLVM_ENABLE_ASSERTIONS=ON \
    -DLIBOMPTARGET_ENABLE_DEBUG=ON \
    -DLLVM_OPTIMIZED_TABLEGEN=ON \
    -DBUILD_SHARED_LIBS=ON \
    -DLLVM_CCACHE_BUILD=${USE_CCACHE} \
    -DLLVM_APPEND_VC_REV=OFF \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
    -DCMAKE_CXX_FLAGS="-Wno-address" \
    -DCMAKE_INSTALL_PREFIX=../llvm-install \
    ../llvm-project/llvm

make -j128 install
```

What would I need to change to specifically rebuild the device libraries, or am I already doing that?

shiltian commented:
@jtramm You are building vanilla LLVM but the device-libs are in [AMD's fork](https://github.com/ROCm/llvm-project/tree/amd-staging/amd/device-libs). You can simply use the LLVM you build to build device-libs. Here is the command I usually use to build device-libs:

```
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DLLVM_ROOT=... -DClang_ROOT=... -B $BUILD_ROOT/device-libs/release -S llvm-project/amd/device-libs
```

jtramm commented:
I don't understand -- are you saying I cannot use vanilla LLVM to compile for AMD anymore, that I need AMD's fork now instead?

shiltian commented:
No, vanilla LLVM doesn't have device-libs. That is only available in AMD's fork. For this particular case, you will have to build device-libs because the device-libs in your system have half-baked values that can cause performance regression. That's why I suggested to simply just remove the attributes from those `/opt/rocm/amdgcn/bitcode/*.bc` files.

You can still build vanilla LLVM, and use that to build device-libs. The only thing you want out of AMD's fork is the device-libs, which sits in `amd/device-libs`.

To make it simple, can you first verify if `amdgpu-flat-work-group-size` or `amdgpu-waves-per-eu` exist in any `/opt/rocm/amdgcn/bitcode/*.bc` files? If there is no, then you don't need to build device-libs. That would be a separate issue.

jtramm commented:
The rocm install on this cluster seems to be in a different location, but the `.bc` files don't appear to have either of those attributes in them:

```
jtramm@amdgpu04:/soft/compilers/rocm/rocm-6.3.0/amdgcn/bitcode$ ls
asanrtl.bc                          oclc_isa_version_1010.bc          oclc_isa_version_10-3-generic.bc  oclc_isa_version_600.bc  oclc_isa_version_805.bc  oclc_isa_version_942.bc
hip.bc                              oclc_isa_version_1011.bc          oclc_isa_version_1100.bc          oclc_isa_version_601.bc  oclc_isa_version_810.bc  oclc_isa_version_9-generic.bc
ockl.bc                             oclc_isa_version_1012.bc          oclc_isa_version_1101.bc          oclc_isa_version_602.bc  oclc_isa_version_900.bc  oclc_unsafe_math_off.bc
oclc_abi_version_400.bc             oclc_isa_version_1013.bc          oclc_isa_version_1102.bc          oclc_isa_version_700.bc  oclc_isa_version_902.bc  oclc_unsafe_math_on.bc
oclc_abi_version_500.bc             oclc_isa_version_10-1-generic.bc  oclc_isa_version_1103.bc          oclc_isa_version_701.bc  oclc_isa_version_904.bc  oclc_wavefrontsize64_off.bc
oclc_abi_version_600.bc             oclc_isa_version_1030.bc          oclc_isa_version_1150.bc          oclc_isa_version_702.bc  oclc_isa_version_906.bc  oclc_wavefrontsize64_on.bc
oclc_correctly_rounded_sqrt_off.bc  oclc_isa_version_1031.bc          oclc_isa_version_1151.bc          oclc_isa_version_703.bc  oclc_isa_version_908.bc  ocml.bc
oclc_correctly_rounded_sqrt_on.bc   oclc_isa_version_1032.bc          oclc_isa_version_1152.bc          oclc_isa_version_704.bc  oclc_isa_version_909.bc  opencl.bc
oclc_daz_opt_off.bc                 oclc_isa_version_1033.bc          oclc_isa_version_11-generic.bc    oclc_isa_version_705.bc  oclc_isa_version_90a.bc
oclc_daz_opt_on.bc                  oclc_isa_version_1034.bc          oclc_isa_version_1200.bc          oclc_isa_version_801.bc  oclc_isa_version_90c.bc
oclc_finite_only_off.bc             oclc_isa_version_1035.bc          oclc_isa_version_1201.bc          oclc_isa_version_802.bc  oclc_isa_version_940.bc
oclc_finite_only_on.bc              oclc_isa_version_1036.bc          oclc_isa_version_12-generic.bc    oclc_isa_version_803.bc  oclc_isa_version_941.bc
jtramm@amdgpu04:/soft/compilers/rocm/rocm-6.3.0/amdgcn/bitcode$ grep amdgpu-flat-work-group-size *.bc
jtramm@amdgpu04:/soft/compilers/rocm/rocm-6.3.0/amdgcn/bitcode$ grep amdgpu-waves-per-eu *.bc
jtramm@amdgpu04:/soft/compilers/rocm/rocm-6.3.0/amdgcn/bitcode$ 
```

jtramm commented:
Ah, it looks like the `.bc` files are binary files. Is there a specific program I need to use to check them for attributes?

shiltian commented:
Can you disassemble the `.bc` to `.ll` and then check again? Directly checking `.bc` will not work because they are compressed/encoded. You can use `/soft/compilers/rocm/rocm-6.3.0/llvm/bin/llvm-dis xxx.bc -o xxx.ll`.
